File: langfuse/04_pipeline/rag_pipeline.py
"""
Lernziele:
- Eine einfache RAG-Pipeline (Retrieval-Augmented Generation) aufbauen
- Keyword-basiertes Retrieval ohne Vektor-Datenbank implementieren
- Jeden Pipeline-Schritt einzeln in Langfuse tracen (Retrieval, Augmentation, Generation)
- Nested Spans für mehrstufige Pipelines nutzen

Voraussetzungen: .env mit LANGFUSE_PUBLIC_KEY, LANGFUSE_SECRET_KEY, LANGFUSE_HOST
                 LM Studio läuft auf http://localhost:1234/v1 mit Qwen3
"""

import os
import sys
from pathlib import Path

from dotenv import load_dotenv
from langfuse import Langfuse
from openai import OpenAI

ENV_PATH = Path(__file__).resolve().parent.parent / ".env"
load_dotenv(ENV_PATH)

# Knowledge base: short text collection about AI
DOCUMENTS = [
    "Künstliche Intelligenz (KI) ist ein Teilgebiet der Informatik, das sich mit der "
    "Automatisierung intelligenten Verhaltens befasst.",
    "Machine Learning ist eine Methode der KI, bei der Algorithmen aus Daten lernen, "
    "anstatt explizit programmiert zu werden.",
    "Deep Learning nutzt künstliche neuronale Netze mit vielen Schichten, um komplexe "
    "Muster in großen Datenmengen zu erkennen.",
    "Natural Language Processing (NLP) ermöglicht es Computern, menschliche Sprache "
    "zu verstehen, zu interpretieren und zu generieren.",
    "Reinforcement Learning ist ein Lernparadigma, bei dem ein Agent durch Versuch "
    "und Irrtum in einer Umgebung optimale Strategien erlernt.",
]


def create_langfuse_client() -> Langfuse:
    """Create and verify a Langfuse client."""
    langfuse = Langfuse()

    if not langfuse.auth_check():
        print(
            "Fehler: Langfuse-Authentifizierung fehlgeschlagen.\n"
            "Prüfe deine .env Datei (siehe .env.example).",
            file=sys.stderr,
        )
        sys.exit(1)

    return langfuse


def create_lm_studio_client() -> OpenAI:
    """Create an OpenAI client pointing to LM Studio."""
    base_url = os.getenv("LM_STUDIO_BASE_URL", "http://localhost:1234/v1")
    return OpenAI(base_url=base_url, api_key="lm-studio")


def retrieve(query: str, documents: list[str], top_k: int = 2) -> list[str]:
    """Simple keyword-based retrieval: score documents by keyword overlap."""
    query_words = set(query.lower().split())

    scored = []
    for doc in documents:
        doc_words = set(doc.lower().split())
        overlap = len(query_words & doc_words)
        scored.append((overlap, doc))

    scored.sort(key=lambda x: x[0], reverse=True)
    return [doc for score, doc in scored[:top_k] if score > 0]


def augment_prompt(query: str, chunks: list[str]) -> str:
    """Build an augmented prompt from the query and retrieved chunks."""
    context = "\n\n".join(f"- {chunk}" for chunk in chunks)
    return (
        "Beantworte die folgende Frage basierend auf dem gegebenen Kontext.\n"
        "Nutze nur Informationen aus dem Kontext. "
        "Wenn der Kontext die Frage nicht beantwortet, sage das ehrlich.\n\n"
        f"Kontext:\n{context}\n\n"
        f"Frage: {query}"
    )


def main() -> None:
    langfuse = create_langfuse_client()
    client = create_lm_studio_client()
    model = os.getenv("LM_STUDIO_MODEL", "qwen3")

    query = "Was ist Machine Learning und wie unterscheidet es sich von Deep Learning?"
    print(f"Frage: {query}\n")

    # -- Root trace for the RAG pipeline --
    trace = langfuse.start_span(
        name="rag-pipeline",
        input={"query": query, "num_documents": len(DOCUMENTS)},
    )

    # -- Step 1: Retrieval --
    retrieval_span = trace.start_span(
        name="retrieval",
        input={"query": query, "top_k": 2},
    )

    chunks = retrieve(query, DOCUMENTS, top_k=2)

    retrieval_span.update(
        output={"chunks": chunks, "num_found": len(chunks)},
    )
    retrieval_span.end()

    print(f"Gefundene Chunks ({len(chunks)}):")
    for i, chunk in enumerate(chunks, 1):
        print(f"  {i}. {chunk[:80]}...")
    print()

    # -- Step 2: Augmentation --
    augmentation_span = trace.start_span(
        name="augmentation",
        input={"query": query, "chunks": chunks},
    )

    augmented_prompt = augment_prompt(query, chunks)

    augmentation_span.update(output={"augmented_prompt": augmented_prompt})
    augmentation_span.end()

    print(f"Augmentierter Prompt:\n{augmented_prompt}\n")

    # -- Step 3: Generation --
    messages = [{"role": "user", "content": augmented_prompt}]
    generation = trace.start_generation(
        name="llm-call",
        model=model,
        input=messages,
        model_parameters={"temperature": 0.3, "max_tokens": 512},
    )

    try:
        response = client.chat.completions.create(
            model=model,
            messages=messages,
            temperature=0.3,
            max_tokens=512,
        )
    except Exception as e:
        generation.update(output=f"Fehler: {e}", level="ERROR")
        generation.end()
        trace.end()
        langfuse.flush()
        print(
            f"Fehler bei der Verbindung zu LM Studio: {e}\n"
            "Stelle sicher, dass LM Studio läuft.",
            file=sys.stderr,
        )
        sys.exit(1)

    answer = response.choices[0].message.content
    usage = response.usage

    usage_details = {}
    if usage:
        usage_details = {
            "input": usage.prompt_tokens,
            "output": usage.completion_tokens,
            "total": usage.total_tokens,
        }

    generation.update(output=answer, usage_details=usage_details)
    generation.end()

    # -- End trace --
    trace.update(
        output={
            "answer": answer,
            "chunks_used": len(chunks),
        }
    )
    trace.end()
    langfuse.flush()

    print(f"Antwort:\n{answer}\n")

    if usage_details:
        print(
            f"Token-Verbrauch: {usage_details.get('input', '?')} input, "
            f"{usage_details.get('output', '?')} output"
        )

    trace_url = langfuse.get_trace_url()
    if trace_url:
        print(f"\nTrace in Langfuse: {trace_url}")

    langfuse.shutdown()


if __name__ == "__main__":
    main()

File: langfuse-ts/src/04_pipeline/ragPipeline.ts
/**
 * Lernziele:
 * - Eine einfache RAG-Pipeline (Retrieval-Augmented Generation) aufbauen
 * - Keyword-basiertes Retrieval ohne Vektor-Datenbank implementieren
 * - Jeden Pipeline-Schritt einzeln in Langfuse tracen (Retrieval, Augmentation, Generation)
 * - Nested Spans für mehrstufige Pipelines nutzen
 *
 * Voraussetzungen: .env mit LANGFUSE_PUBLIC_KEY, LANGFUSE_SECRET_KEY, LANGFUSE_HOST
 *                  LM Studio läuft auf http://localhost:1234/v1 mit Qwen3
 */

import Langfuse from "langfuse";
import OpenAI from "openai";
import "dotenv/config";

// Knowledge base: short text collection about AI
const DOCUMENTS: string[] = [
  "Künstliche Intelligenz (KI) ist ein Teilgebiet der Informatik, das sich mit der " +
    "Automatisierung intelligenten Verhaltens befasst.",
  "Machine Learning ist eine Methode der KI, bei der Algorithmen aus Daten lernen, " +
    "anstatt explizit programmiert zu werden.",
  "Deep Learning nutzt künstliche neuronale Netze mit vielen Schichten, um komplexe " +
    "Muster in großen Datenmengen zu erkennen.",
  "Natural Language Processing (NLP) ermöglicht es Computern, menschliche Sprache " +
    "zu verstehen, zu interpretieren und zu generieren.",
  "Reinforcement Learning ist ein Lernparadigma, bei dem ein Agent durch Versuch " +
    "und Irrtum in einer Umgebung optimale Strategien erlernt.",
];

function createLangfuseClient(): Langfuse {
  return new Langfuse();
}

function createLmStudioClient(): OpenAI {
  const baseURL = process.env.LM_STUDIO_BASE_URL ?? "http://localhost:1234/v1";
  return new OpenAI({ baseURL, apiKey: "lm-studio" });
}

// Simple keyword-based retrieval: score documents by keyword overlap
function retrieve(
  query: string,
  documents: string[],
  topK: number = 2
): string[] {
  const queryWords = new Set(query.toLowerCase().split(/\s+/));

  const scored: [number, string][] = documents.map((doc) => {
    const docWords = new Set(doc.toLowerCase().split(/\s+/));
    let overlap = 0;
    for (const word of queryWords) {
      if (docWords.has(word)) overlap++;
    }
    return [overlap, doc];
  });

  scored.sort((a, b) => b[0] - a[0]);
  return scored
    .filter(([score]) => score > 0)
    .slice(0, topK)
    .map(([, doc]) => doc);
}

// Build an augmented prompt from the query and retrieved chunks
function augmentPrompt(query: string, chunks: string[]): string {
  const context = chunks.map((chunk) => `- ${chunk}`).join("\n\n");
  return (
    "Beantworte die folgende Frage basierend auf dem gegebenen Kontext.\n" +
    "Nutze nur Informationen aus dem Kontext. " +
    "Wenn der Kontext die Frage nicht beantwortet, sage das ehrlich.\n\n" +
    `Kontext:\n${context}\n\n` +
    `Frage: ${query}`
  );
}

async function main(): Promise<void> {
  const langfuse = createLangfuseClient();
  const client = createLmStudioClient();
  const model = process.env.LM_STUDIO_MODEL ?? "qwen3";

  const query =
    "Was ist Machine Learning und wie unterscheidet es sich von Deep Learning?";
  console.log(`Frage: ${query}\n`);

  // -- Root trace for the RAG pipeline --
  const trace = langfuse.trace({
    name: "rag-pipeline",
    input: { query, num_documents: DOCUMENTS.length },
  });

  // -- Step 1: Retrieval --
  const retrievalSpan = trace.span({
    name: "retrieval",
    input: { query, top_k: 2 },
  });

  const chunks = retrieve(query, DOCUMENTS, 2);

  retrievalSpan.end({
    output: { chunks, num_found: chunks.length },
  });

  console.log(`Gefundene Chunks (${chunks.length}):`);
  chunks.forEach((chunk, i) => {
    console.log(`  ${i + 1}. ${chunk.substring(0, 80)}...`);
  });
  console.log();

  // -- Step 2: Augmentation --
  const augmentationSpan = trace.span({
    name: "augmentation",
    input: { query, chunks },
  });

  const augmentedPrompt = augmentPrompt(query, chunks);

  augmentationSpan.end({ output: { augmented_prompt: augmentedPrompt } });

  console.log(`Augmentierter Prompt:\n${augmentedPrompt}\n`);

  // -- Step 3: Generation --
  const messages: OpenAI.Chat.ChatCompletionMessageParam[] = [
    { role: "user", content: augmentedPrompt },
  ];

  const generation = trace.generation({
    name: "llm-call",
    model,
    input: messages,
    modelParameters: { temperature: 0.3, max_tokens: 512 },
  });

  let response: OpenAI.Chat.ChatCompletion;
  try {
    response = await client.chat.completions.create({
      model,
      messages,
      temperature: 0.3,
      max_tokens: 512,
    });
  } catch (e) {
    generation.end({ output: `Fehler: ${e}` });
    await langfuse.flushAsync();
    console.error(
      `Fehler bei der Verbindung zu LM Studio: ${e}\n` +
        "Stelle sicher, dass LM Studio läuft."
    );
    process.exit(1);
  }

  const answer = response.choices[0].message.content!;
  const usage = response.usage;

  const usageDetails: Record<string, number> = {};
  if (usage) {
    usageDetails.input = usage.prompt_tokens;
    usageDetails.output = usage.completion_tokens;
    usageDetails.total = usage.total_tokens;
  }

  generation.end({ output: answer, usageDetails });

  // -- End trace --
  trace.update({
    output: {
      answer,
      chunks_used: chunks.length,
    },
  });
  await langfuse.flushAsync();

  console.log(`Antwort:\n${answer}\n`);

  if (Object.keys(usageDetails).length > 0) {
    console.log(
      `Token-Verbrauch: ${usageDetails.input ?? "?"} input, ` +
        `${usageDetails.output ?? "?"} output`
    );
  }

  const traceUrl = trace.getTraceUrl();
  console.log(`\nTrace in Langfuse: ${traceUrl}`);

  await langfuse.shutdownAsync();
}

main();
